service/frontend/workflow_handler.go

```diff
@@ -4749,6 +4750,357 @@ func (wh *WorkflowHandler) StartBatchOperation(
+		identity = op.UpdateActivityOptionsOperation.GetIdentity()
+		operation := &batchpb.BatchOperation_UpdateActivityOptionsOperation{
+			UpdateActivityOptionsOperation: &batchpb.BatchOperationUpdateActivityOptions{
+				Identity: op.UpdateActivityOptionsOperation.GetIdentity(),
+			},
+		}
+		if op.UpdateActivityOptionsOperation == nil {
+			return nil, serviceerror.NewInvalidArgument("update activity options operation is not set")
+		}
+		if op.UpdateActivityOptionsOperation.GetActivityOptions() != nil && op.UpdateActivityOptionsOperation.GetRestoreOriginal() {
+			return nil, serviceerror.NewInvalidArgument("cannot set both activity options and restore original")
+		}
+		if op.UpdateActivityOptionsOperation.GetActivityOptions() == nil && !op.UpdateActivityOptionsOperation.GetRestoreOriginal() {
+			return nil, serviceerror.NewInvalidArgument("Either activity type must be set, or restore original should be set to true")
+		}
+
+		switch a := op.UpdateActivityOptionsOperation.GetActivity().(type) {
+		case *batchpb.BatchOperationUpdateActivityOptions_Type:
+			if len(a.Type) == 0 {
+				return nil, serviceerror.NewInvalidArgument("Either activity type must be set, or match all should be set to true")
+			}
+			operation.UpdateActivityOptionsOperation.Activity = &batchpb.BatchOperationUpdateActivityOptions_Type{
+				Type: a.Type,
+			}
+		case *batchpb.BatchOperationUpdateActivityOptions_MatchAll:
+			if !a.MatchAll {
+				return nil, serviceerror.NewInvalidArgument("Either activity type must be set, or match all should be set to true")
+			}
+			operation.UpdateActivityOptionsOperation.Activity = &batchpb.BatchOperationUpdateActivityOptions_MatchAll{
+				MatchAll: true,
+			}
+		}
+
+		operation.UpdateActivityOptionsOperation.RestoreOriginal = op.UpdateActivityOptionsOperation.GetRestoreOriginal()
+		operation.UpdateActivityOptionsOperation.Identity = op.UpdateActivityOptionsOperation.GetIdentity()
+		if updateMask := op.UpdateActivityOptionsOperation.GetUpdateMask(); updateMask != nil {
+			operation.UpdateActivityOptionsOperation.UpdateMask = updateMask
+		}
+		if ao := op.UpdateActivityOptionsOperation.GetActivityOptions(); ao != nil {
+			operation.UpdateActivityOptionsOperation.ActivityOptions = &activitypb.ActivityOptions{
+				ScheduleToStartTimeout: ao.ScheduleToStartTimeout,
+				ScheduleToCloseTimeout: ao.ScheduleToCloseTimeout,
+				StartToCloseTimeout:    ao.StartToCloseTimeout,
+				HeartbeatTimeout:       ao.HeartbeatTimeout,
+			}
+			if rp := ao.RetryPolicy; rp != nil {
+				operation.UpdateActivityOptionsOperation.ActivityOptions.RetryPolicy = &commonpb.RetryPolicy{
+					InitialInterval:        rp.InitialInterval,
+					MaximumInterval:        rp.MaximumInterval,
+					BackoffCoefficient:     rp.BackoffCoefficient,
+					NonRetryableErrorTypes: rp.NonRetryableErrorTypes,
+					MaximumAttempts:        rp.MaximumAttempts,
+				}
+			}
+			operation.UpdateActivityOptionsOperation.ActivityOptions.TaskQueue = ao.TaskQueue
+		}
```

Comment:
Same as you did elsewhere but now I'm wondering if we can't just pass in the original request to the batch job and be done with it...

```suggestion
		ao := op.UpdateActivityOptionsOperation.GetActivityOptions()
```